File: src/codec/types/inventory.cpp
/// InventoryAction codec implementation.

#include "endweave/codec/types/inventory.h"
#include "endweave/codec/types/enums.h"

namespace endweave {

std::expected<InventoryAction, ReadError> read_inventory_action(PacketReader& reader) {
    auto source_type = reader.read_uvarint();
    if (!source_type) return std::unexpected(source_type.error());

    InventoryAction action;
    action.source_type = *source_type;

    if (*source_type == static_cast<std::uint32_t>(InventorySourceType::ContainerInventory) ||
        *source_type == static_cast<std::uint32_t>(InventorySourceType::NonImplementedFeatureTODO)) {
        auto wid = reader.read_varint();
        if (!wid) return std::unexpected(wid.error());
        action.window_id = *wid;
    } else if (*source_type == static_cast<std::uint32_t>(InventorySourceType::WorldInteraction)) {
        auto flags = reader.read_uvarint();
        if (!flags) return std::unexpected(flags.error());
        action.source_flags = *flags;
    }

    auto slot = reader.read_uvarint();
    if (!slot) return std::unexpected(slot.error());
    action.slot = *slot;

    auto old_item = read_item_instance(reader);
    if (!old_item) return std::unexpected(old_item.error());
    action.old_item = std::move(*old_item);

    auto new_item = read_item_instance(reader);
    if (!new_item) return std::unexpected(new_item.error());
    action.new_item = std::move(*new_item);

    return action;
}

void write_inventory_action(PacketWriter& writer, const InventoryAction& action) {
    writer.write_uvarint(action.source_type);

    if (action.source_type == static_cast<std::uint32_t>(InventorySourceType::ContainerInventory) ||
        action.source_type == static_cast<std::uint32_t>(InventorySourceType::NonImplementedFeatureTODO)) {
        writer.write_varint(action.window_id.value_or(0));
    } else if (action.source_type == static_cast<std::uint32_t>(InventorySourceType::WorldInteraction)) {
        writer.write_uvarint(action.source_flags.value_or(0));
    }

    writer.write_uvarint(action.slot);
    write_item_instance(writer, action.old_item);
    write_item_instance(writer, action.new_item);
}

template <> auto PacketReader::read<inventory_action>() -> std::expected<InventoryAction, ReadError> {
    return read_inventory_action(*this);
}
template <> void PacketWriter::write<inventory_action>(const InventoryAction& val) {
    write_inventory_action(*this, val);
}

} // namespace endweave

File: src/endstone_endweave/codec/types/inventory.py
"""Inventory-related compound types (InventoryAction)."""

from dataclasses import dataclass

from ..reader import PacketReader
from ..writer import PacketWriter
from .enums import InventorySourceType
from .item import ITEM_INSTANCE, ItemInstance
from .primitives import UVAR_INT, VAR_INT, Type


@dataclass
class InventoryAction:
    """A single InventoryAction entry in InventoryTransactionPacket.

    Attributes:
        source_type: Source type ID (0=Container, 1=Global, 2=World, 3=Creative).
        window_id: Window ID (only for Container/NonImplementedFeature source types).
        source_flags: Source flags (only for WorldInteraction source type).
        slot: Inventory slot index.
        old_item: Item before the action.
        new_item: Item after the action.
    """

    source_type: int
    window_id: int | None
    source_flags: int | None
    slot: int
    old_item: ItemInstance
    new_item: ItemInstance


class _InventoryActionType(Type["InventoryAction"]):
    """InventoryAction compound type with conditional fields."""

    def read(self, reader: PacketReader) -> InventoryAction:
        source_type = UVAR_INT.read(reader)  # SourceType
        window_id: int | None = None
        source_flags: int | None = None
        if source_type in (InventorySourceType.CONTAINER_INVENTORY, InventorySourceType.NON_IMPLEMENTED_FEATURE_TODO):
            window_id = VAR_INT.read(reader)  # WindowID
        elif source_type == InventorySourceType.WORLD_INTERACTION:
            source_flags = UVAR_INT.read(reader)  # SourceFlags
        slot = UVAR_INT.read(reader)  # InventorySlot
        old_item = ITEM_INSTANCE.read(reader)  # OldItem
        new_item = ITEM_INSTANCE.read(reader)  # NewItem
        return InventoryAction(source_type, window_id, source_flags, slot, old_item, new_item)

    def write(self, writer: PacketWriter, value: InventoryAction) -> None:
        UVAR_INT.write(writer, value.source_type)  # SourceType
        if value.source_type in (
            InventorySourceType.CONTAINER_INVENTORY,
            InventorySourceType.NON_IMPLEMENTED_FEATURE_TODO,
        ):
            assert value.window_id is not None
            VAR_INT.write(writer, value.window_id)  # WindowID
        elif value.source_type == InventorySourceType.WORLD_INTERACTION:
            assert value.source_flags is not None
            UVAR_INT.write(writer, value.source_flags)  # SourceFlags
        UVAR_INT.write(writer, value.slot)  # InventorySlot
        ITEM_INSTANCE.write(writer, value.old_item)  # OldItem
        ITEM_INSTANCE.write(writer, value.new_item)  # NewItem


INVENTORY_ACTION = _InventoryActionType()
